aiida/engine/processes/workchains/restart.py

```diff
@@ -320,8 +321,17 @@ def results(self) -> Optional['ExitCode']:
             return self.exit_codes.ERROR_MAXIMUM_ITERATIONS_EXCEEDED  # pylint: disable=no-member
 
         self.report(f'work chain completed after {self.ctx.iteration} iterations')
+        self._attach_outputs(node)
+        return None
+
+    def _attach_outputs(self, node) -> Mapping[str, orm.Node]:
```

Comment:
Do we want to make this a hidden method, since developers should call it in their process handlers?

Reply:
I think this is fine, because the process handlers are instance methods of the `WorkChain`, so they will have access to this method just fine. They will do `self._attach_outputs`. And I don't think this method should be called from outside of the workchain (which would only really happen in unit tests anyway, I think).